### PR TITLE
Implement string id for tool qualities. 

### PR DESCRIPTION
**Pull request by BevapDin**
What the title says. Also adds a check for valid tool quality of vehicle parts and uses `generic_factory` for storing the quality data.

### CODE REVIEW

File: src/iexamine.cpp [hunk 1]
```diff
@@ -653,7 +653,7 @@ void iexamine::cardreader(player &p, const tripoint &examp)
 
 void iexamine::rubble(player &p, const tripoint &examp)
 {
-    bool has_digging_tool = p.has_quality( "DIG", 2 );
+    bool has_digging_tool = p.has_quality( quality_id( "DIG" ), 2 );
```
Comment (mugling) [inline src/iexamine.cpp]: Out of curiosity why is `string_id<T>()` constructor marked `explicit`. Presumably this is why you require syntax such as this?

**Comment by Coolthulhu** [comment]
My compiler:
```
Ubuntu clang version 3.6.0-2ubuntu1~trusty1 (tags/RELEASE_360/final) (based on LLVM 3.6.0)
```
Says this:
```
src/generic_factory.h:191:9: error: constructor for 'generic_factory<quality>' must explicitly initialize the const member 'dummy_obj'
        generic_factory( const std::string &type_name, const std::string &id_member_name = "id",
```

**Comment by Coolthulhu** [comment]
I think I see why: `quality` struct has `bool was_loaded = false;`, but other fields unspecified. Clang seems to (sometimes?) refuse creating a default constructor in such a case or marks it as explicit.

File: src/iexamine.cpp (hunk 1 above)
Comment (BevapDin) [inline src/iexamine.cpp]: ------------------------------------
Mostly to encourage few explicit instantiating of it. Creating a temporary like here is fine if it's done only once (for that value) in a compilation unit. If it's done several times, one should use a constant, either function local like in `iuse::trap`, or global. This allows spellchecking by the compiler.

However, this line here is not the desired result. Ideally the condition would be `p.has( quality_id( "DIG" ), 2 )`, a function named `has` (or similar), overloaded for various id types. That gives a less redundant, but type safe interface: `if( has( some_trait_id ) && has( some_effect_id, bodypart ) && has( itype_id( "..." ) ) )`.

And it allows to *easily find all literal ids in the code.

@Coolthulhu I fixed it, I should probably use clang more often.

**Pull request merged**